ENH Expose verbose_feature_names_out in make_union

Ethan0456 opened the pull request:
### Reference Issues/PRs
Fixes #30364

### What does this implement/fix? Explain your changes.
This PR exposes the `verbose_feature_names_out` argument in the `make_union` function. It allows users to control feature name uniqueness in the `FeatureUnion` by enabling or disabling automatic name prefixing, which helps in managing feature name collisions more effectively.

### Any other comments?
No additional comments.

Ethan0456 committed:
ENH Expose verbose_feature_names_out in make_union

Signed-off-by: Abhijeetsingh Meena <abhijeet040403@gmail.com>

Ethan0456 committed:
STYLE Fix linting

Signed-off-by: Abhijeetsingh Meena <abhijeet040403@gmail.com>

thomasjpfan reviewed (commented):
Thank you for the PR @Ethan0456 !

- Can you add a test to `sklearn/tests/test_pipeline.py`?
- Also add a whats new in `doc/whats_new/upcoming_changes/sklearn.pipeline/30406.enhancement.rst`. You can follow the pattern that directory to see how to create new entry.

Ethan0456 committed:
TST DOC Add test cases and documentation for verbose_feature_names_out in make_union

- Added a test case to validate the `verbose_feature_names_out` argument in `make_union`, including handling of duplicate feature names.
- Added an enhancement entry in `doc/whats_new/upcoming_changes/sklearn.pipeline/30406.enhancement.rst`.

References #30364

Signed-off-by: Abhijeetsingh Meena <abhijeet040403@gmail.com>

Ethan0456 committed:
STYLE Fix linting

Signed-off-by: Abhijeetsingh Meena <abhijeet040403@gmail.com>

thomasjpfan commented on sklearn/tests/test_pipeline.py:
```diff
@@ -603,6 +603,67 @@ def test_make_union_kwargs():
+    return mock
+
+
+def test_make_union_verbose_feature_names_out_enabled():
```
Since, we already have similar test for `FeatureUnion`. I think it's enough to test that `make_union` sets `verbose_feature_names_out` correctly. Something like this:

```python
def test_make_union_passes_verbose_feature_names_out():
    union = make_union(..., verbose_feature_names_out=False)
    assert not union.verbose_feature_names_out
```

Ethan0456 replied:
Hi @thomasjpfan,

I've simplified the test case in my recent commit. Could you please review it when you get a chance?

Thanks!

Ethan0456 committed:
TST Simplify test for `make_union` verbose_feature_names_out flag

Signed-off-by: Abhijeetsingh Meena <abhijeet040403@gmail.com>

thomasjpfan reviewed (approved):
LGTM

OmarManzoor commented on sklearn/tests/test_pipeline.py:
```diff
@@ -603,6 +603,14 @@ def test_make_union_kwargs():
+    mock = Transf()
+
+    union = make_union(pca, mock, verbose_feature_names_out=False)
+    assert not union.verbose_feature_names_out
```
Would it be possible to also have a case where `verbose_feature_names_out=True` and we can verify that the feature names are as expected?

Ethan0456 replied:
Hi @OmarManzoor,

I've added the case to check `verbose_feature_names_out=True` in my recent commit. Thanks for providing feedback, I look forward to your thoughts when you have time.

Thanks!

OmarManzoor reviewed (commented):
Thanks for the PR @Ethan0456

Ethan0456 committed:
TST Test make_union with verbose_feature_names_out set to True

OmarManzoor reviewed (approved):
LGTM. Thanks @Ethan0456
